domains/client/block-builder/src/custom_api.rs

```diff
@@ -0,0 +1,384 @@
+//! Custom API that is efficient to collect storage roots.
+
+use codec::{Codec, Decode, Encode};
+use hash_db::{HashDB, Hasher, Prefix};
+use sc_client_api::{backend, ExecutorProvider, StateBackend};
+use sp_core::offchain::OffchainOverlayedChange;
+use sp_core::traits::{CallContext, CodeExecutor, RuntimeCode};
+use sp_inherents::InherentData;
+use sp_runtime::traits::{Block as BlockT, HashingFor, NumberFor};
+use sp_runtime::{ApplyExtrinsicResult, ExtrinsicInclusionMode, TransactionOutcome};
+use sp_state_machine::backend::AsTrieBackend;
+use sp_state_machine::{
+    BackendTransaction, DBValue, IndexOperation, OverlayedChanges, StateMachine, StorageChanges,
+    StorageKey, StorageValue, TrieBackend, TrieBackendBuilder, TrieBackendStorage,
+};
+use std::collections::HashMap;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+type TrieBackendStorageFor<State, Block> =
+    <State as StateBackend<HashingFor<Block>>>::TrieBackendStorage;
+
+pub(crate) type TrieDeltaBackendFor<'a, State, Block> = TrieBackend<
+    DeltaBackend<'a, TrieBackendStorageFor<State, Block>, HashingFor<Block>>,
+    HashingFor<Block>,
+>;
+
+struct MappedStorageChanges<H: Hasher> {
+    /// All changes to the main storage.
+    ///
+    /// A value of `None` means that it was deleted.
+    pub main_storage_changes: HashMap<StorageKey, Option<StorageValue>>,
+    /// All changes to the child storages.
+    pub child_storage_changes: HashMap<StorageKey, HashMap<StorageKey, Option<StorageValue>>>,
+    /// Offchain state changes to write to the offchain database.
+    pub offchain_storage_changes: HashMap<(Vec<u8>, Vec<u8>), OffchainOverlayedChange>,
+    /// A transaction for the backend that contains all changes from
+    /// [`main_storage_changes`](StorageChanges::main_storage_changes) and from
+    /// [`child_storage_changes`](StorageChanges::child_storage_changes).
```

Comment:
```suggestion
    /// [`child_storage_changes`](StorageChanges::child_storage_changes) and from
```

Reply:
Sorry for not explaining what I meant. There is an incomplete sentence about offchain storage changes on the line below, can it be removed, or connected to this sentence?

```suggestion
    /// [`child_storage_changes`](StorageChanges::child_storage_changes), but excluding
```

Reply:
ahh I see. Will update